### PR TITLE
ftsender支持无队列并发发送

### PR DESCRIPTION
**Pull request by Yezersky**
## Changes
 
  - [x] ftsender支持无队列并发发送，新增ft_strategy选项`concurrent`
  - [x] 增加direct queue，用来支持无队列并发，起到goroutine间（`FtSender.Send`和`FtSender.sendFromQueue`之间）通讯的作用。direct queue实际上并不是队列，只是实现了队列接口。
  - [x] 移除memory queue
 
## Reviewers

  - [x] @wonderflow  please review

## Checklist
   
   - [x] Rebased/mergable
   - [x] Tests pass
   - [ ] CHANGELOG.md updated
   - [ ] Jira issue/task done

### CODE REVIEW

File: sender/fault_tolerant_sender.go [hunk 1]
```diff
@@ -28,7 +29,7 @@ const (
 	KeyFtSaveLogPath       = "ft_save_log_path" // disk queue 数据日志路径
 	KeyFtWriteLimit        = "ft_write_limit"   // 写入速度限制，单位MB
 	KeyFtStrategy          = "ft_strategy"      // ft 的策略
-	KeyFtProcs             = "ft_procs"         // ft并发数，当always_save 策略时启用
+	KeyFtProcs             = "ft_procs"         // ft并发数，当always_save或concurrent策略时启用
```
Comment (wonderflow) [inline sender/fault_tolerant_sender.go]: 在sender_model那个地方也加上相应的选项

https://github.com/qiniu/logkit/blob/develop/sender/rest_sender_models.go

File: sender/fault_tolerant_sender_test.go [hunk 2]
```diff
@@ -192,3 +192,37 @@ func TestFtChannelFullSender(t *testing.T) {
...
+	mp := conf.MapConf{}
+	mp[KeyFtSaveLogPath] = tmpDir
+	mp[KeyFtStrategy] = KeyFtStrategyConcurrent
+	mp[KeyFtProcs] = "3"
+	defer os.RemoveAll(tmpDir)
+	fts, err := NewFtSender(s, mp)
+	assert.NoError(t, err)
+	datas := []Data{
+		{"ab": "abcccc"},
+		{"ab": "E18111:"},
```
Comment (wonderflow) [inline sender/fault_tolerant_sender_test.go]: 这个E18111 不是会出错的吗

File: sender/fault_tolerant_sender.go [hunk 3]
```diff
@@ -82,14 +85,21 @@ func NewFtSender(sender Sender, conf conf.MapConf) (*FtSender, error) {
 	syncEvery, _ := conf.GetIntOr(KeyFtSyncEvery, DefaultFtSyncEvery)
 	writeLimit, _ := conf.GetIntOr(KeyFtWriteLimit, defaultWriteLimit)
 	strategy, _ := conf.GetStringOr(KeyFtStrategy, KeyFtStrategyAlwaysSave)
+	switch strategy {
+	case KeyFtStrategyAlwaysSave:
+	case KeyFtStrategyBackupOnly:
+	case KeyFtStrategyConcurrent:
```
Comment (wonderflow) [inline sender/fault_tolerant_sender.go]: 上面三行用逗号连接，变成一行好看一些

File: sender/fault_tolerant_sender_test.go (hunk 2 above)
Comment (Yezersky) [reply sender/fault_tolerant_sender_test.go]: 不会出错，用的是mock sender。我还是改下以免引起歧义。

**Comment by Yezersky** [comment]
理论上减少了很多开销，之后会在单元测试里加一组Benchmark。

File: sender/rest_sender_models.go [hunk 4]
```diff
@@ -179,7 +179,7 @@ var ModeKeyOptions = map[string][]utils.Option{
 			ChooseOnly:   false,
 			Default:      "",
 			DefaultNoUse: false,
-			Description:  "发送并发数量(磁盘管道或内存管道 always_save 模式生效)(ft_procs)",
+			Description:  "发送并发数量(磁盘管道或内存管道 always_save 或 concurrent 模式生效)(ft_procs)",
```
Comment (wonderflow) [inline sender/rest_sender_models.go]: KeyFtStrategy

搜下这个哈，还是有遗漏

**Comment by Yezersky** [comment]
BenchmarkFtSenderConcurrentDirect-8   	  300000	      4425 ns/op
--- BENCH: BenchmarkFtSenderConcurrentDirect-8
	fault_tolerant_sender_test.go:283: Benchmark.N: 1
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 1
	fault_tolerant_sender_test.go:283: Benchmark.N: 100
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 100
	fault_tolerant_sender_test.go:283: Benchmark.N: 10000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 10000
	fault_tolerant_sender_test.go:283: Benchmark.N: 300000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 300000
BenchmarkFtSenderConcurrentDisk-8     	   20000	     77574 ns/op
--- BENCH: BenchmarkFtSenderConcurrentDisk-8
	fault_tolerant_sender_test.go:283: Benchmark.N: 1
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 1
	fault_tolerant_sender_test.go:283: Benchmark.N: 100
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 100
	fault_tolerant_sender_test.go:283: Benchmark.N: 10000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 10000
	fault_tolerant_sender_test.go:283: Benchmark.N: 20000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 20000
BenchmarkFtSenderConcurrentMemory-8   	   10000	    461803 ns/op
--- BENCH: BenchmarkFtSenderConcurrentMemory-8
	fault_tolerant_sender_test.go:283: Benchmark.N: 1
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 1
	fault_tolerant_sender_test.go:283: Benchmark.N: 100
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 64
	fault_tolerant_sender_test.go:283: Benchmark.N: 10000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 9953

**Comment by wonderflow** [comment]
LGTM

**Comment by Yezersky** [comment]
之前的Benchmark中FtSenderConcurrentMemory结果异常是MockSender导致，修正这个问题后的Benchmark结果如下：

```
BenchmarkFtSenderConcurrentDirect-8   	  300000	      4438 ns/op
--- BENCH: BenchmarkFtSenderConcurrentDirect-8
	fault_tolerant_sender_test.go:283: Benchmark.N: 1
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 1
	fault_tolerant_sender_test.go:283: Benchmark.N: 100
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 100
	fault_tolerant_sender_test.go:283: Benchmark.N: 10000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 10000
	fault_tolerant_sender_test.go:283: Benchmark.N: 300000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 300000
BenchmarkFtSenderConcurrentDisk-8     	   20000	     79477 ns/op
--- BENCH: BenchmarkFtSenderConcurrentDisk-8
	fault_tolerant_sender_test.go:283: Benchmark.N: 1
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 1
	fault_tolerant_sender_test.go:283: Benchmark.N: 100
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 100
	fault_tolerant_sender_test.go:283: Benchmark.N: 10000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 10000
	fault_tolerant_sender_test.go:283: Benchmark.N: 20000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 20000
BenchmarkFtSenderConcurrentMemory-8   	  200000	      6351 ns/op
--- BENCH: BenchmarkFtSenderConcurrentMemory-8
	fault_tolerant_sender_test.go:283: Benchmark.N: 1
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 0
	fault_tolerant_sender_test.go:283: Benchmark.N: 100
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 59
	fault_tolerant_sender_test.go:283: Benchmark.N: 10000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 9904
	fault_tolerant_sender_test.go:283: Benchmark.N: 200000
	fault_tolerant_sender_test.go:284: MockSender.SendCount: 199902
```

**Comment by SunRunAway** [comment]
https://godoc.org/golang.org/x/perf/cmd/benchstat

建议后面用这个工具贴出benchmark对比结果

**Pull request merged**